src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java

```diff
@@ -90,9 +94,38 @@ public interface JavaNioAccess {
      * (for the duration of the I/O operation) by acquiring the session.
      * Null is returned if the buffer has no scope, or acquiring is not
      * required to guarantee safety.
+     * {@snippet lang = java:
+     * var handler = acquireSession(buffer, async);
+     * try {
+     *     performOperation(buffer);
+     * } finally {
+     *     if (handler != null) {
+     *         handler.run();
+     *     }
+     * }
+     *}
      */
     Runnable acquireSession(Buffer buffer, boolean async);
 
+    /**
+     * Used by operations to make a buffer's session non-closeable
+     * (for the duration of the operation) by acquiring the session.
+     * The buffers scope is returned if it has one, otherwise {@code null} is returned.
+     * {@snippet lang = java:
+     * var guard = acquireScopeOrNull(buffer);
+     * try {
+     *     performOperation(buffer);
+     * } finally {
+     *     releaseScope(buffer, guard);
+     * }
+     *}
+     *
+     * @see #releaseScope(Buffer, MemorySessionImpl)
+     */
+    MemorySessionImpl acquireScopeOrNull(Buffer buffer);
```

Comment:
I think this looks better - but naming-wise it's still a bit problematic. This method really acquires the underlying session, not a scope. And also, here we have `OrNull`, but the already existing `acquireSession` also has a similar semantics w.r.t. null.

I suggest to rename:

* `acquireSession` -> `acquireSessionAsRunnable`
* `acquireScopeOrNull` -> `acquireSession`
* `releaseScope` -> `releaseSession`

Also, once we have `acquire/releaseSession`, it is not clear to me that we still need `acquireSessionAsRunnable` in the JavaNIOAccess class - it seems like you can create the Runnable where it's required (probably IOUtil), simply by using straight acquire/release.

Reply:
The name "scope" was used in anticipation of the new proposed Java 20 naming. But I can change it back to session again. We could always rename later.